truffle/src/main/java/org/jruby/truffle/nodes/core/ArrayNodes.java

```diff
@@ -2615,159 +2618,143 @@ public CallTarget getCallTarget() {
+                return new RaiseException(getContext().getCoreLibrary().argumentError("X outside of string", this));
+            } catch (CantCompressNegativeException e) {
+                return new RaiseException(getContext().getCoreLibrary().argumentError("can't compress negative numbers", this));
+            } catch (RangeException e) {
+                return new RaiseException(getContext().getCoreLibrary().rangeError(e.getMessage(), this));
+            } catch (CantConvertException e) {
+                return new RaiseException(getContext().getCoreLibrary().typeError(e.getMessage(), this));
+            }
+        }
+
+        private RubyString finishPack(ByteList format, PackResult result) {
+            final RubyString string = getContext().makeString(new ByteList(result.getOutput(), 0, result.getOutputLength()));
+
+            if (format.length() == 0) {
+                string.forceEncoding(USASCIIEncoding.INSTANCE);
+            } else {
+                switch (result.getEncoding()) {
+                    case DEFAULT:
+                    case ASCII_8BIT:
+                        break;
+                    case US_ASCII:
+                        string.forceEncoding(USASCIIEncoding.INSTANCE);
+                        break;
+                    case UTF_8:
+                        string.forceEncoding(UTF8Encoding.INSTANCE);
+                        break;
+                    default:
+                        throw new UnsupportedOperationException();
+                }
             }
-        }
 
-        @Specialization(guards = "!isRubyString(format)")
-        public RubyString pack(VirtualFrame frame, RubyArray array, Object format) {
-            // TODO CS 1-Mar-15 sloppy until I can get @CreateCast to work
+            if (result.isTainted()) {
+                if (taintNode == null) {
+                    CompilerDirectives.transferToInterpreter();
+                    taintNode = insert(TaintNodeFactory.create(getContext(), getEncapsulatingSourceSection(), null));
+                }
 
-            if (toStringNode == null) {
-                CompilerDirectives.transferToInterpreter();
-                toStringNode = insert(DispatchHeadNodeFactory.createMethodCall(getContext(), MissingBehavior.RETURN_MISSING));
+                taintNode.executeTaint(string);
             }
 
-            final Object result = toStringNode.call(frame, format, "to_str", null);
-
-            if (result == DispatchNode.MISSING) {
-                throw new RaiseException(getContext().getCoreLibrary().typeErrorNoImplicitConversion(format, "String", this));
-            }
+            return string;
+        }
 
-            if (result instanceof RubyString) {
-                return pack(frame, array, (RubyString) result);
-            }
+        @Specialization
+        public Object pack(VirtualFrame frame, RubyArray array, boolean format) {
+            return ruby(frame, "raise TypeError");
```

Comment:
I'm not such a big fan of raising exception with the ruby() helper.
Notably raise is implemented rather late and the overhead is likely significant, while the gains minimal IMO.
But of course in this case I suppose a good ToStr node would do it instead of the extra specializations.